ocaml/typing/jkind.ml

```diff
@@ -261,57 +302,71 @@ module Const = struct
 
   let get_legacy_layout
-      { layout; modes_upper_bounds = _; externality_upper_bound } :
-      Layout.Const.Legacy.t =
-    match layout, externality_upper_bound with
-    | Any, _ -> Any
-    | Sort Value, Internal -> Value
-    | Sort Value, External64 -> Immediate64
-    | Sort Value, External -> Immediate
-    | Sort Void, _ -> Void
-    | Sort Float64, _ -> Float64
-    | Sort Float32, _ -> Float32
-    | Sort Word, _ -> Word
-    | Sort Bits32, _ -> Bits32
-    | Sort Bits64, _ -> Bits64
+      { layout;
+        modes_upper_bounds = _;
+        externality_upper_bound;
+        nullability_upper_bound
+      } : Layout.Const.Legacy.t =
+    match layout, externality_upper_bound, nullability_upper_bound with
+    | Any, _, Or_null -> Any
+    | Any, _, Non_null -> Any_non_null
+    (* CR layouts v3.0: support [Immediate(64)_or_null]. *)
+    | Sort Value, _, Or_null -> Value_or_null
+    | Sort Value, Internal, Non_null -> Value
+    | Sort Value, External64, Non_null -> Immediate64
+    | Sort Value, External, Non_null -> Immediate
+    | Sort Void, _, _ -> Void
+    | Sort Float64, _, _ -> Float64
+    | Sort Float32, _, _ -> Float32
+    | Sort Word, _, _ -> Word
+    | Sort Bits32, _, _ -> Bits32
+    | Sort Bits64, _, _ -> Bits64
 
   let equal
       { layout = lay1;
         modes_upper_bounds = modes1;
-        externality_upper_bound = ext1
+        externality_upper_bound = ext1;
+        nullability_upper_bound = null1
       }
       { layout = lay2;
         modes_upper_bounds = modes2;
-        externality_upper_bound = ext2
+        externality_upper_bound = ext2;
+        nullability_upper_bound = null2
       } =
     Layout.Const.equal lay1 lay2
     && Modes.equal modes1 modes2
     && Externality.equal ext1 ext2
+    && Nullability.equal null1 null2
 
   let sub
       { layout = lay1;
         modes_upper_bounds = modes1;
-        externality_upper_bound = ext1
+        externality_upper_bound = ext1;
+        nullability_upper_bound = null1
       }
       { layout = lay2;
         modes_upper_bounds = modes2;
-        externality_upper_bound = ext2
+        externality_upper_bound = ext2;
+        nullability_upper_bound = null2
       } =
     Misc.Le_result.combine_list
       [ Layout.Const.sub lay1 lay2;
         Modes.less_or_equal modes1 modes2;
-        Externality.less_or_equal ext1 ext2 ]
+        Externality.less_or_equal ext1 ext2;
+        Nullability.less_or_equal null1 null2 ]
 
   let not_mode_crossing layout =
     { layout;
       modes_upper_bounds = Modes.max;
-      externality_upper_bound = Externality.max
+      externality_upper_bound = Externality.max;
+      nullability_upper_bound = Nullability.max
     }
 
   let mode_crossing layout =
     { layout;
       modes_upper_bounds = Modes.min;
-      externality_upper_bound = Externality.min
+      externality_upper_bound = Externality.min;
+      nullability_upper_bound = Nullability.min
```

Comment:
I personally find it confusing to include this in `mode_crossing`. It's not really a mode in the same sense as the others. You've convinced me that it can be seen as a bound, but I would not expect to have the `mode_crossing` function affect nullability. Can you move this out from this function?

Reply:
No -- that is a mode. It's deep!

Reply:
Fixed.